components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +341,101 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
+        let max_month = Self::months_in_year_cached(cache);
+        if month > max_month {
+            return Err(CalendarError::Overflow {
+                field: "month",
+                max: max_month as usize,
+            });
+        }
+
+        let max_day = Self::days_in_month(month, cache.new_year);
+        if day > max_day {
+            return Err(CalendarError::Overflow {
+                field: "day",
+                max: max_day as usize,
+            });
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+
+        Ok(ArithmeticDate::<C>::new_unchecked(year, month, day))
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(&self.1)
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also takes a `ChineseBasedCache` argument.
+    fn months_in_year_cached(cache: &ChineseBasedCache) -> u8 {
+        if cache.leap_month.is_some() {
+            13
+        } else {
+            12
+        }
+    }
+
+    /// Calls `days_in_month` on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_month_inner(&self) -> u8 {
+        Self::days_in_month(self.0.month, self.1.new_year)
+    }
+
+    /// Returns the number of days in the given `month` after the given `new_year`.
+    fn days_in_month(month: u8, new_year: RataDie) -> u8 {
+        let approx = new_year + ((month - 1) as i64 * 29);
+        let prev_new_moon = Self::new_moon_before((approx + 15).as_moment());
+        let next_new_moon = Self::new_moon_on_or_after((approx + 15).as_moment());
+        let result = (next_new_moon - prev_new_moon) as u8;
+        debug_assert!(result == 29 || result == 30);
+        result
+    }
+
+    /// Calls day_in_year_cached on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_year_inner(&self) -> u16 {
+        let next_new_year = Self::new_year_on_or_before_fixed_date(self.1.new_year + 400, None).0;
+        Self::days_in_year(self.1.new_year, next_new_year)
+    }
+
+    /// Returns the number of day in the given year bounds
+    fn days_in_year(prev_new_year: RataDie, next_new_year: RataDie) -> u16 {
+        let result = next_new_year - prev_new_year;
+        debug_assert!(
+            ((u16::MIN as i64)..=(u16::MAX as i64)).contains(&result),
+            "Days in year should be in range of u16."
+        );
+        result as u16
+    }
+
+    /// Compute a `ChineseBasedCache` from a ChineseBased year
+    pub(crate) fn compute_cache(year: i32) -> ChineseBasedCache {
+        let mid_year = Self::fixed_mid_year_from_year(year);
+        let prior_solstice = Self::winter_solstice_on_or_before(mid_year);
+        let new_year = Self::new_year_on_or_before_fixed_date(mid_year, Some(prior_solstice)).0;
+        let is_leap_year = Self::new_year_is_leap_year(new_year, None);
```

Comment:
Can you retain the solstice from `new_year_on_or_before_fixed_date` and pass it into `new_year_is_leap_year`?

Reply:
Yes, done